server/cache.go

```diff
@@ -0,0 +1,115 @@
+package server
+
+import (
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/itzg/mc-router/mcproto"
+	"github.com/sirupsen/logrus"
+
+	mcpinger "github.com/Raqbit/mc-pinger"
+)
+
+type StatusCache struct {
+	mu    sync.RWMutex
+	cache map[string]*mcproto.StatusResponse // key: serverAddress
+}
+
+func NewStatusCache() *StatusCache {
+	return &StatusCache{
+		cache: make(map[string]*mcproto.StatusResponse),
+	}
+}
+
+func (sc *StatusCache) Get(serverAddress string) (*mcproto.StatusResponse, bool) {
+	sc.mu.RLock()
+	defer sc.mu.RUnlock()
+	status, ok := sc.cache[serverAddress]
+	if !ok {
+		return nil, false
+	}
+	return status, true
+}
+
+func (sc *StatusCache) Set(serverAddress string, status *mcproto.StatusResponse) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	sc.cache[serverAddress] = status
+}
+
+func (sc *StatusCache) Delete(serverAddress string) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	delete(sc.cache, serverAddress)
+}
+
+func (sc *StatusCache) updateAll(getBackends func() map[string]string) {
```

Comment:
Nice encapsulation of passing in the getter function! Since it gets passed in a couple of places, I wonder if it could be provided once to `NewStatusCache`?